connectors/src/lib/data_sources.ts

```diff
@@ -229,43 +232,101 @@ async function _updateDocumentParentsField({
   }
 }
 
-const MAX_SECTION_PREFIX_LENGTH = 192;
+// allows for 4 full prefixes before hitting half of the max chunk size (approx.
+// 256 chars for 512 token chunks)
+const MAX_PREFIX_TOKENS = EMBEDDING_CONFIG.max_chunk_size / 8;
+// Limit on chars to avoid tokenizing too much text uselessly on documents with
+// large prefixes. The final truncating will rely on MAX_PREFIX_TOKENS so this
+// limit can be large and should be large to avoid underusing prexfixes
+const MAX_PREFIX_CHARS = MAX_PREFIX_TOKENS * 8;
 
 // The role of this function is to create a prefix from an arbitrary long string. The prefix
 // provided will not be augmented with `\n`, so it should include appropriate carriage return. If
-// the prefix is too long (>256 chars), it will be truncated. The remained will be returned as
+// the prefix is too long (> MAX_PREFIX_TOKENS), it will be truncated. The remained will be returned as
 // content of the resulting section.
-export function renderPrefixSection(
+export async function renderPrefixSection(
+  dataSourceConfig: DataSourceConfig,
   prefix: string | null
-): CoreAPIDataSourceDocumentSection {
+): Promise<CoreAPIDataSourceDocumentSection> {
   if (!prefix || !prefix.trim()) {
     return {
       prefix: null,
       content: null,
       sections: [],
     };
   }
-  if (prefix.length > MAX_SECTION_PREFIX_LENGTH) {
+  let finalPrefix = prefix;
+  if (prefix.length > MAX_PREFIX_CHARS) {
+    finalPrefix = prefix.substring(0, MAX_PREFIX_CHARS);
+  }
+  const tokens = (await tokenize(finalPrefix, dataSourceConfig)).map(
+    (token) => token[1]
+  );
+  if (tokens.length <= MAX_PREFIX_TOKENS) {
     return {
-      prefix: prefix.substring(0, MAX_SECTION_PREFIX_LENGTH) + "...\n",
-      content: `... ${prefix.substring(MAX_SECTION_PREFIX_LENGTH)}`,
+      prefix: prefix,
+      content: null,
       sections: [],
     };
   }
+  const prefixTextLength =
+    tokens.slice(0, MAX_PREFIX_TOKENS).reduce((acc, t) => acc + t.length, 0) -
```

Comment:
As per IRL, let's hope the assumption that no character is lost when tokenizing is true 🤞

Reply:
Good point. From my tests + what I read on BPE, it is true.
If there are rare cases in which a char or 2 is lost, it will have almost no consequence (a prefix a few chars shorter than what it could but no data loss)